File: Medium/Search_A_2D_Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        // want to find correct row first then can binary search in row

        int rowIndex = 0;
        int columnIndex = 0;
        int rowSize = matrix.length;
        int columnSize = matrix[0].length;

        // binary search to find correct row
        int left = 0;
        int right = rowSize - 1;
        int mid = 0;
        while(left <= right){
            
            mid = left + (right - left) / 2;
            if(mid+1 > rowSize-1){
               rowIndex = rowSize-1;
               break; 
            }
            if(matrix[mid][0] <= target && matrix[mid+1][0] > target){
                rowIndex = mid;
                break;
            }
            else if(matrix[mid][0] < target){
                left = mid + 1;
            }
            else if(matrix[mid][0] > target){
                right = mid - 1;
            }
        }

        // binary search to find correct column
        left = 0;
        right = columnSize - 1;
        while(left <= right){
            
            mid = left + (right - left) / 2;
            
            if(matrix[rowIndex][mid] ==target){
                return true;
            }
            else if(matrix[rowIndex][mid] < target){
                left = mid + 1;
            }
            else if(matrix[rowIndex][mid] > target){
                right = mid - 1;
            }
        }

        



        return false;
    }
}

File: Medium/Search_A_2D_Matrix.cs
public class Solution {
    public bool SearchMatrix(int[][] matrix, int target) {
        // first find the row, then find the value in that row. Use binary search for both
        int rowSize = matrix.Length;
        int colSize = matrix[0].Length;
        

        // find the targetRow first
        int targetRow = -1;
        int top = 0;
        int bottom = rowSize-1;
        int mid = 0;
        while (top <= bottom){
            mid = top + (bottom - top) / 2;
            int leftVal = matrix[mid][0];
            int rightVal = matrix[mid][colSize-1];
            

            if (target == leftVal || target == rightVal){
                return true;
            }
            if (target > leftVal && target < rightVal){
                targetRow = mid;
                break;
            }
            if (target < leftVal){
                bottom = mid-1;
            }
            else if (target > rightVal){
                top = mid+1;
            }

        }
        if(targetRow == -1){
            return false;
        }

        // find the target in the row
        int left = 0;
        int right = colSize-1;
        Console.WriteLine(targetRow);
        while(left <= right){
            mid = left + (right - left) / 2;
            if(matrix[targetRow][mid] == target){
                return true;
            }
            else if(target < matrix[targetRow][mid]){
                right = mid-1;
            }
            else{
                left = mid+1;
            }
        }




        return false;

    }
}